Improve performance by using parameters in queries [5]

<!--
Please read the Contributing guidelines (https://github.com/hzi-braunschweig/SORMAS-Project/blob/development/docs/CONTRIBUTING.md) before submitting an issue. You don't have to remove this comment or any other comment from this issue as they will automatically be hidden.
-->
### Problem Description
<!-- Mandatory -->

I have noticed that in `PersonService::buildSimilarityCriteriaFilter` `cb.literal(name)` is set directly, not a parameter.
Same goes for the other filters. 
Having a reusable query definition could be beneficial to improve performance.

_Originally posted by @HolgerReiseVSys in https://github.com/hzi-braunschweig/SORMAS-Project/issues/3598#issuecomment-735841830_

Finding in https://github.com/hzi-braunschweig/SORMAS-Project/issues/3915#issuecomment-1234240264 : The current approaches with literals use parameter bindings except for number values.

### Proposed Change
<!-- Mandatory -->

~~Replace all usages of `cb.literal(name)` by using a parameter, as explained here: https://www.baeldung.com/jpa-query-parameters#criteria-query-parameters.
The only exception to this are constants - e.g. "configuration" values of psql functions that are called like `cb.literal("[^0-9]")`.~~

- Use `<property name="hibernate.criteria.literal_handling_mode" value="bind" />` to force alywas binding.

### Acceptance Criteria
<!-- Optional -->

- [x] Make sure that by default bind mode is used.

### Implementation Details
<!-- Optional -->

### Additional Information
<!-- Optional -->

One collateral finding on test-de from now:
```bash
Caused by: javax.ejb.TransactionRolledbackLocalException: Exception thrown from bean: java.lang.IllegalArgumentException: Parameter value [251119] did not match expected type [de.symeda.sormas.backend.caze.Case (n/a)]
        at com.sun.ejb.containers.EJBContainerTransactionManager.checkExceptionClientTx(EJBContainerTransactionManager.java:615)
        at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:485)
        at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4601)
        at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2134)
        ... 112 more
Caused by: java.lang.IllegalArgumentException: Parameter value [251119] did not match expected type [de.symeda.sormas.backend.caze.Case (n/a)]
        at org.hibernate.query.spi.QueryParameterBindingValidator.validate(QueryParameterBindingValidator.java:54)
        at org.hibernate.query.spi.QueryParameterBindingValidator.validate(QueryParameterBindingValidator.java:27)
        at org.hibernate.query.internal.QueryParameterBindingImpl.validate(QueryParameterBindingImpl.java:90)
        at org.hibernate.query.internal.QueryParameterBindingImpl.setBindValue(QueryParameterBindingImpl.java:55)
        at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:501)
        at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:122)
        at org.hibernate.query.criteria.internal.compile.CriteriaCompiler$1$1.bind(CriteriaCompiler.java:135)
        at org.hibernate.query.criteria.internal.CriteriaQueryImpl$1.buildCompiledQuery(CriteriaQueryImpl.java:364)
        at org.hibernate.query.criteria.internal.compile.CriteriaCompiler.compile(CriteriaCompiler.java:165)
        at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:774)
        at org.hibernate.internal.AbstractSessionImpl.createQuery(AbstractSessionImpl.java:23)
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper.createQuery(EntityManagerWrapper.java:498)
        at de.symeda.sormas.backend.contact.ContactFacadeEjb.getContactCountsByCasesForDashboard(ContactFacadeEjb.java:1302)
...
        at de.symeda.sormas.api.contact.__ContactFacade_Remote_DynamicStub.getContactCountsByCasesForDashboard(de/symeda/sormas/api/contact/__ContactFacade_Remote_DynamicStub.java)
        at de.symeda.sormas.api.contact._ContactFacade_Wrapper.getContactCountsByCasesForDashboard(de/symeda/sormas/api/contact/_ContactFacade_Wrapper.java)
        at de.symeda.sormas.ui.dashboard.contacts.ContactsDashboardView.updateCaseCountsAndSourceCasesLabels(ContactsDashboardView.java:269)
        at de.symeda.sormas.ui.dashboard.contacts.ContactsDashboardView.refreshDashboard(ContactsDashboardView.java:475)
        at de.symeda.sormas.ui.dashboard.contacts.ContactsDashboardView.enter(ContactsDashboardView.java:144)
```

--- a/sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java
+++ b/sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java
@@ -1296,7 +1296,7 @@ public class ContactFacadeEjb
 			Root<Contact> contact2 = cq2.from(Contact.class);
 			cq2.groupBy(contact2.get(Contact.CAZE));
 
-			cq2.where(contact2.get(Contact.CAZE).in(caseIds));
+			cq2.where(contact2.get(Contact.CAZE).get(Case.ID).in(caseIds));
 			cq2.select(cb.count(contact2.get(Contact.ID)));
 
 			List<Long> caseContactCounts = em.createQuery(cq2).getResultList();

--- a/sormas-backend/src/test/java/de/symeda/sormas/backend/contact/ContactFacadeEjbTest.java
+++ b/sormas-backend/src/test/java/de/symeda/sormas/backend/contact/ContactFacadeEjbTest.java
@@ -18,6 +18,7 @@
 package de.symeda.sormas.backend.contact;
 
 import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
@@ -28,7 +29,6 @@ import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.Mockito.when;
 
@@ -957,12 +957,27 @@ public class ContactFacadeEjbTest extends AbstractBeanTest {
 
 		List<Long> ids;
 
-		// test with some random id: returns 0,0,0
+		// 0. test with some random id: returns 0,0,0
 		ids = Arrays.asList(5555L);
 		int[] result = getContactFacade().getContactCountsByCasesForDashboard(ids);
 		assertThat(result[0], equalTo(0));
 		assertThat(result[1], equalTo(0));
 		assertThat(result[2], equalTo(0));
+
+		RDCFEntities rdcfEntities = creator.createRDCFEntities("Region", "District", "Community", "Facility");
+		RDCF rdcf = new RDCF(rdcfEntities);
+		UserDto user = useNationalAdminLogin();
+
+		// 1. Test with one contact
+		PersonDto cazePerson = creator.createPerson("Case", "Person");
+		CaseDataDto caze = createCaze(user, cazePerson, rdcfEntities);
+		ContactDto contact = createContact(user, caze, rdcf);
+		ids = Arrays.asList(getContactService().getByUuid(contact.getUuid()).getId());
+		result = getContactFacade().getContactCountsByCasesForDashboard(ids);
+
+		assertThat(result[0], equalTo(1));
+		assertThat(result[1], equalTo(1));
+		assertThat(result[2], equalTo(1));
 	}
 
 	@Test

